### PR TITLE
Correctly construct outputs when spending internal notes.

### PR DESCRIPTION
**Pull request by nuttycom**
This builds atop #658

### CODE REVIEW

File: zcash_client_backend/src/data_api/wallet.rs [hunk 1]
```diff
@@ -303,19 +303,27 @@ where
...
     // Create the transaction
     let mut builder = Builder::new(params.clone(), height);
     for selected in spendable_notes {
-        let from = {
+        // If the diversifier corresponds to that for the internal change address, use that
+        // address to construct the note, otherwise derive the external address corresponding
+        // to the diversifier for the selected note.
+        //
+        // FIXME: This only works so long as the wallet never chooses anything but the default
+        // diversifier index for creating a change address.
```
Comment (str4d) [inline zcash_client_backend/src/data_api/wallet.rs:316]: I agree that this shouldn't occur in general (for wallets following ZIP 316), but I'd prefer if we could check internal vs external reliably. And I've thought of a hacky way to do it 🙂
```rust
let note = {
    let external_note = dfvk.fvk()
        .vk
        .to_payment_address(selected.diversifier)
        .unwrap()
        .create_note(selected.note_value.into(), selected.rseed);
    let internal_note = dfvk.internal()
        .fvk()
        .vk
        .to_payment_address(selected.diversifier)
        .unwrap()
        .create_note(selected.note_value.into(), selected.rseed);

    let expected_root = selected.witness.root();
    if expected_root == merkle_path.root(external_note.cmu()) {
        Ok(external_note)
    else if expected_root == merkle_path.root(internal_note.cmu()) {
        Ok(internal_note)
    } else {
        Err(TODO) 
    }
}?;
```

**Review by str4d — CHANGES_REQUESTED** [review]
Reviewed f3f31dab1cea2a3490e02001c5461737f52095f1.

File: zcash_primitives/CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,63 @@ and this library adheres to Rust's notion of
...
+    - `Builder::transparent_outputs()`
+    - `Builder::sapling_inputs()`
+    - `Builder::sapling_outputs()`
+- `zcash_primitives::transaction::fees` a new module containing abstractions
+  and types related to fee calculations.
+  - `TransactionBalance` A type representing the unspent balance of transaction inputs,
+    allocated separately to change and fee portions.
+  - `FeeRule` a function that computes the fee required for a transaction given
+    inputs and outputs to the transaction.
+  - `Infalliable` an uninhabited error type for conditions where no error can occur.
```
Comment (pacu) [inline zcash_primitives/CHANGELOG.md:29]:
```suggestion
  - `Infallible` an uninhabited error type for conditions where no error can occur.
```

**Review by pacu — APPROVED** [review]
utACK

File: zcash_primitives/src/sapling.rs [hunk 3]
```diff
@@ -517,6 +525,12 @@ impl Note {
             )),
         }
     }
+
+    pub fn commitment(&self) -> Node {
```
Comment (str4d) [inline zcash_primitives/src/sapling.rs:529]: Needs documentation.

**Review by str4d — APPROVED** [review]
utACK cdfca848eae42e7b3396da558859695cf36767c3 after changelog updates.

**Review by str4d — APPROVED** [review]
Re-utACK 981d45e9661445ecaef757c497d66af1f249ef3c. Changes over last review are only to documentation.

**Pull request merged**